src/main/java/org/purescript/psi/PSModule.kt

```diff
@@ -181,6 +226,35 @@ class PSModule(node: ASTNode) :
             PSExportedData::class.java
         )
 
+    /**
+     * @return the [PSDataConstructor] elements that this module exports,
+     * both directly and through re-exported modules
+     */
+    val exportedDataConstructors: List<PSDataConstructor>
+        get() {
+            val explicitlyExportedItems = exportList?.exportedItems
+                ?: return dataConstructors
+
+            val exportedDataConstructors = mutableListOf<PSDataConstructor>()
+
+            for (exportedData in explicitlyExportedItems.filterIsInstance<PSExportedData>()) {
+                if (exportedData.exportsAll) {
+                    exportedData.dataDeclaration?.dataConstructors
+                        ?.mapTo(exportedDataConstructors) { it }
+                } else {
+                    exportedData.dataMembers
+                        .mapNotNull { it.reference.resolve() }
+                        .filterIsInstanceTo(exportedDataConstructors)
+                }
+            }
+
+            explicitlyExportedItems.filterIsInstance<PSExportedModule>()
+                .mapNotNull { it.importDeclaration }
+                .flatMapTo(exportedDataConstructors) { it.importedDataConstructors }
+
+            return exportedDataConstructors
+        }
+
```

Comment:
Due to the way data and newtype constructors are resolved in the export list, we cannot reuse functionality for the other exported/imported items. We probably want to find an abstraction to reduce the code duplication here and in the import declarations.